core/arch/arch.c

```diff
@@ -3355,45 +3355,59 @@ dr_mcontext_to_priv_mcontext(priv_mcontext_t *dst, dr_mcontext_t *src)
      * if we append to dr_mcontext_t in the future we'll need
      * to check src->size here.
      */
-    if (src->size != sizeof(dr_mcontext_t) &&
-        src->size != sizeof(dr_mcontext_t) - sizeof(dr_opmask_t))
+    if (src->size > sizeof(dr_mcontext_t))
         return false;
-    if (TESTALL(DR_MC_ALL, src->flags))
+    if (TEST(DR_MC_ALL, src->flags) &&
+        IF_X86_ELSE(src->size > offsetof(priv_mcontext_t, opmask), true)) {
         *dst = *(priv_mcontext_t *)(&MCXT_FIRST_REG_FIELD(src));
-    else {
-        if (TEST(DR_MC_INTEGER, src->flags)) {
+    } else {
+        if (TESTANY(DR_MC_INTEGER | DR_MC_ALL, src->flags)) {
             /* xsp is in the middle of the mcxt, so we save dst->xsp here and
              * restore it later so we can use one memcpy for DR_MC_INTEGER.
              */
             reg_t save_xsp = dst->xsp;
-            memcpy(&MCXT_FIRST_REG_FIELD(dst), &MCXT_FIRST_REG_FIELD(src),
-                   /* end of the mcxt integer gpr */
-                   offsetof(priv_mcontext_t, IF_X86_ELSE(xflags, pc)));
+            if (src->size >= offsetof(priv_mcontext_t, IF_X86_ELSE(xflags, pc))) {
```

Comment:
bug: The size is of dr_mcontext_t, not priv_mcontext_t.  The extra fields need to be accounted for.

Ditto below.

Reply:
I should have seen dr_mcontext_init. I assumed the size is set only accounting for priv_mcontext_t. But this wouldn't make sense. Thanks. Fixed.